src/spec/helpers/iterate.cr

```diff
@@ -0,0 +1,94 @@
+module Spec::Methods
+  # Spec helper for generic iteration methods which tests both yielding and
+  # iterator overloads.
+  #
+  # This helper creates two spec examples named *description* with suffixes
+  # `" yielding"` and `" iterator"`.
+  # The yielding example calls *method* with a block and expects the iteration
+  # elements to be yielded to the block. The iterator example calls *method*
+  # without a block and expects it to return an `Iterator` which it then consumes.
+  #
+  # The iterated elements are collected in an array and compared to *expected*,
+  # ensuring type-equality of the elements.
+  #
+  # By default, both examples make sure that the iteration is finished after
+  # iterating all elements from *expected*. If the iteration is infinite,
+  # passing `infinite: true` skips that check and allows to test a finite sample
+  # of an infinite iteration.
+  #
+  # ```cr
+  # require "spec/helpers/iterate"
+  #
+  # it_iterates "Array#each", [1, 2, 3], (1..3).each
+  # it_iterates "infinite #cycle", [1, 2, 3, 1, 2, 3, 1], (1..3).cycle, infinite: true
+  # ```
+  #
+  # If the iteration elements are tuples (i.e. multiple values), the yielding
+  # variant by default only catches the first value because of the block argument
+  # mechanics. Passing `tuple: true` ensures all yielded arguments are collected
```

Comment:
tuple: true is a bit cryptic. Maybe something like:

yielded_args: :single (default)
yielded_args: :all

Reply:
Reminded me of this discussion
https://github.com/crystal-lang/crystal/pull/10109#issuecomment-753475897

But actually here I'd more likely support a boolean. It's a bit similar to `reuse: true`

Reply:
maybe a boolean `single_block_arg = true`? Or `tuple_block_arg = false`?